### ISSUE TITLE
decode('UTF-8') is not correct.

### ISSUE THREAD

**Issue by spotrh** [rpm-software-management/rpmlint#202]
As filed here: https://bugzilla.redhat.com/show_bug.cgi?id=1699789

RPM Python binding started to warn on calling decode() on non-byte strings to
in order to make the Python 3 binding Unicode-correct. It seems that
rpmlint-1.10-22.fc31.noarch uses the bogus decode() method:

$ rpmlint -f perl-Socket.rpmlintrc perl-Socket-2.029-2.fc31.src.rpm 
/usr/share/rpmlint/Pkg.py:168: UnicodeWarning: decode() called on unicode
string, see https://bugzilla.redhat.com/show_bug.cgi?id=1693751
  s.decode('UTF-8')
1 packages and 0 specfiles checked; 0 errors, 0 warnings.

It happens when calling rpmlint on any RPM package. I have
python3-rpm-4.14.2.1-7.fc31.x86_64.

**Comment by scarabeusiv** [comment]
Hi, looking at the code it is partialy bogus. The code in 1.10:
https://github.com/rpm-software-management/rpmlint/blob/rpmlint-1.10/Pkg.py#L166

Currently it is still present and it is used just to check if it is utf8 or not so we even want the exception:
https://github.com/rpm-software-management/rpmlint/blob/master/rpmlint/pkg.py#L144

Honestly that detection still needs to be redone anyway and it is not much used in the code:
https://github.com/rpm-software-management/rpmlint/search?q=is_utf8&unscoped_q=is_utf8

**Comment by hroncok** [comment]
So until the change in Python RPM bindings, this was bytes, so we tried to decode it and if ti works, we are good.

Since the Python RPM bindings chnage, the decode is esentialy a no-op, so this returns True and detects nothing. This API change also leads to (on 1.11):

```
Check that rpmlint executes with no unexpected errors
...in default locale
sh: /usr/bin/python: No such file or directory
/builddir/build/BUILD/rpmlint-rpmlint-1.11/Pkg.py:176: UnicodeWarning: decode() called on unicode string, see https://bugzilla.redhat.com/show_bug.cgi?id=1693751
  s.decode('UTF-8')
/builddir/build/BUILD/rpmlint-rpmlint-1.11/Pkg.py:176: UnicodeWarning: decode() called on unicode string, see https://bugzilla.redhat.com/show_bug.cgi?id=1693751
  s.decode('UTF-8')
Traceback (most recent call last):
  File "./rpmlint", line 376, in <module>
    main()
  File "./rpmlint", line 166, in main
    runChecks(pkg)
  File "./rpmlint", line 224, in runChecks
    check.check(pkg)
  File "/builddir/build/BUILD/rpmlint-rpmlint-1.11/AbstractCheck.py", line 51, in check
    return self.check_binary(pkg)
  File "/builddir/build/BUILD/rpmlint-rpmlint-1.11/PostCheck.py", line 103, in check_binary
    tag[2], prereq)
  File "/builddir/build/BUILD/rpmlint-rpmlint-1.11/PostCheck.py", line 170, in check_aux
    if incorrect_shell_script(prog, script):
  File "/builddir/build/BUILD/rpmlint-rpmlint-1.11/PostCheck.py", line 63, in incorrect_shell_script
    return check_syntax_script(prog, '-n', shellscript)
  File "/builddir/build/BUILD/rpmlint-rpmlint-1.11/PostCheck.py", line 77, in check_syntax_script
    tmpfile.write(script)
TypeError: a bytes-like object is required, not 'str'
```

**Comment by hroncok** [comment]
To the original is_utf8_bytestr problem: In the new API, the thing is UTF-8 decoded, errors-surrogateescaped Unicode string such as this:

```pycon
>>> 'Hrončok'.encode('iso-8859-2').decode('utf-8', errors='surrogateescape')
'Hron\udce8ok'
>>> 'Hron\udce8ok'[4]
'\udce8'
```

In order to check whether an unicode string has surrogates in it, [there is an RFE](https://bugs.python.org/issue25269). We can try to *encode* it:

```
>>> 'Hron\udce8ok'.encode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'utf-8' codec can't encode character '\udce8' in position 4: surrogates not allowed
```

So in order if an RPM returned header is indeed UTF-8 OK we would need to do something like this:

```python
def utf8able(header):
    try:
        header.encode('utf-8')
    except UnicodeEncodeError:
        return False
    except AttributeError:
        try:
            header.decode('utf-8')
        except UnicodeDecodeError:
            return False
        else:
            return True
    else:
        return True
```

Which is... lovely.

**Comment by scarabeusiv** [comment]
Yea that looks lovely, and probably the shortest way to fix it :(

**Comment by spotrh** [comment]
Worth noting that this issue is now causing rpmlint to fail to run in Fedora rawhide:

https://bugzilla.redhat.com/show_bug.cgi?id=1722868

**Comment by hroncok** [comment]
OK, I wanted to work on this, but I have zero idea how to test rpmlint from a git checkout.

This is what I've tried:

```
[rpmlint (master)]$ python3 -m venv __venv__ --system-site-packages
[rpmlint (master)]$ . __venv__/bin/activate
(__venv__) [rpmlint (master)]$ pip install -e .
Obtaining file://.../rpmlint
Requirement already satisfied: pyxdg in /usr/lib/python3.7/site-packages (from rpmlint==2.0.0) (0.26)
Requirement already satisfied: ConfigObj in /usr/lib/python3.7/site-packages (from rpmlint==2.0.0) (5.0.6)
Installing collected packages: rpmlint
  Running setup.py develop for rpmlint
Successfully installed rpmlint
You are using pip version 19.0.3, however version 19.1.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
(__venv__) [rpmlint (master)]$ which rpmlint
~/.../rpmlint/__venv__/bin/rpmlint
(__venv__) [rpmlint (master)]$ rpmlint --version
rpmlint version 2.0.0
(__venv__) [rpmlint (master)]$ rpmlint rpm
sh: /usr/bin/python: No such file or directory
0 packages and 0 specfiles checked; 0 errors, 21 warnings.
Traceback (most recent call last):
  File ".../rpmlint/__venv__/bin/rpmlint", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File ".../rpmlint/build/_scripts/rpmlint", line 314, in <module>
    main()
  File ".../rpmlint/build/_scripts/rpmlint", line 150, in main
    runChecks(pkg)
  File ".../rpmlint/build/_scripts/rpmlint", line 206, in runChecks
    check.check(pkg)
  File ".../rpmlint/rpmlint/checks/TagsCheck.py", line 108, in check
    if self.configuration['Packager'] and \
AttributeError: 'TagsCheck' object has no attribute 'configuration'
```

**Comment by hroncok** [comment]
I've commented that part out and I was able to come up with https://github.com/rpm-software-management/rpmlint/pull/212

**Comment by hroncok** [comment]
with rpm 4.15, I also had to change https://github.com/rpm-software-management/rpmlint/blob/0565ecdaa95dad2a2ae63e9bda521deb2334f910/rpmlint/checks/PostCheck.py#L59

to

```python
    tmpfile = os.fdopen(tmpfd, 'w')
```

Since the script is now unicode.
